### PR TITLE
Xiao/validate signature last

### PR DESCRIPTION
**Pull request by ciaozhang**
Introduce a boolean flag(validateSignatureLast) to control the token validation order:
validate issuer before validate signature and validate payload

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/Validators.cs [hunk 1]
```diff
@@ -249,26 +249,13 @@ internal static string ValidateIssuer(string issuer, SecurityToken securityToken
                 throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidIssuerException(LogMessages.IDX10204)
                 { InvalidIssuer = issuer });
 
-            if (configuration != null)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Validators.cs:252]: Why was this removed?

File: src/Microsoft.IdentityModel.Tokens/Validators.cs [hunk 2]
```diff
@@ -286,8 +273,17 @@ internal static string ValidateIssuer(string issuer, SecurityToken securityToken
                 }
             }
 
+            if (configuration != null)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/Validators.cs]: We should check configuration first, in this path we expect to match configuration.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 3]
```diff
@@ -1243,14 +1287,44 @@ private TokenValidationResult ValidateJWE(JsonWebToken jwtToken, string decrypte
...
+                if (validatedToken == null)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10505, token)));
+
+                if (!(validatedToken is JsonWebToken validatedJsonWebToken))
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10506, LogHelper.MarkAsNonPII(typeof(JsonWebToken)), LogHelper.MarkAsNonPII(validatedToken.GetType()), token)));
+
+                return validatedJsonWebToken;
+            }
+
+            return null;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: If we get here, this is a bug and will result in a null ref in ValidateTokenPayload.
Should we throw here, in JwtSecurityTokenHandler we check for null in ValidateTokenPayload.

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 4]
```diff
@@ -1004,13 +1037,63 @@ private ClaimsPrincipal ValidateToken(string token, JwtSecurityToken outerToken,
...
+
+                return jwtToken;
+            }
+            else
+            {
+                return ReadJwtToken(token);
+            }
+        }
+
+            /// <summary>
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: Is the indentation off in these comments?

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 5]
```diff
@@ -1004,13 +1037,63 @@ private ClaimsPrincipal ValidateToken(string token, JwtSecurityToken outerToken,
...
+                if (validatedJwtToken == null)
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10505, token)));
+
+                if (!(validatedJwtToken is JwtSecurityToken validatedJwt))
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10506, LogHelper.MarkAsNonPII(typeof(JwtSecurityToken)), LogHelper.MarkAsNonPII(validatedJwtToken.GetType()), token)));
+
+                return validatedJwt;
+            }
+
+            return null;
```
Comment (RojaEnnam) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:1066]: Consider checking is at least one delegate is present at method start and changing to if..else(line 1062)

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 3 above)
Comment (RojaEnnam) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: Consider checking if at least one delegate is present at method start and changing to if..else(line 1304)

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 6]
```diff
@@ -1204,8 +1204,51 @@ private TokenValidationResult ValidateJWS(string token, TokenValidationParameter
...
+                        jsonWebToken = new JsonWebToken(token);
+                    }
+
+                    if (validationParameters.ValidateSignatureLast)
+                    {
+                        tokenValidationResult = ValidateTokenPayload(jsonWebToken, validationParameters, configuration);
+                        if (tokenValidationResult.IsValid)
+                        {
+                            tokenValidationResult.SecurityToken = ValidateSignature(token, jsonWebToken, validationParameters, configuration);
+                            Validators.ValidateIssuerSecurityKey(tokenValidationResult.SecurityToken.SigningKey, jsonWebToken, validationParameters);
```
Comment (RojaEnnam) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: Consider coupling ValidateSignature and ValidateIssuerSecurityKey. so that we can avoid checking the flag in ValidateTokenPayload.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 7]
```diff
@@ -1204,8 +1204,51 @@ private TokenValidationResult ValidateJWS(string token, TokenValidationParameter
         {
             try
             {
-                var jsonWebToken = ValidateSignature(token, validationParameters, configuration);
-                return ValidateTokenPayload(jsonWebToken, validationParameters, configuration);
+                JsonWebToken jsonWebToken;
+                TokenValidationResult tokenValidationResult;
+                if (validationParameters.SignatureValidator != null || validationParameters.SignatureValidatorUsingConfiguration != null)
+                {
+                    return ValidateTokenPayload(
```
Comment (RojaEnnam) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: Shouldn't we check ValidateSignatureLast here as well?

**Review by brentschmaltz — APPROVED** [review]

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs (hunk 4 above)
Comment (ciaozhang) [reply src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: Fixed the indentation

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 7 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: SignatureValidator and SignatureValidatorUsingConfiguration will return a JsonWebToken to validate the payload. Set tvp.ValidateSignatureLast = false here.

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs (hunk 5 above)
Comment (ciaozhang) [reply src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:1066]: resolved

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 3 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: resolved

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 6 above)
Comment (dannybtsai) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: I think Roja's suggestion makes sense. Having to check the `validationParameters.ValidateSignatureLast` in two methods is more error prone. Consider wrapping the two methods:

```
        private static JsonWebToken ValidateSignatureAndIssuerSecurityKey(string token, JsonWebToken jsonWebToken, TokenValidationParameters validationParameters, BaseConfiguration configuration)
        {
            var validatedToken = ValidateSignature(token, jsonWebToken, validationParameters, configuration);
            Validators.ValidateIssuerSecurityKey(validatedToken.SigningKey, jsonWebToken, validationParameters);

            return validatedToken;
        }
```

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 7 above)
Comment (dannybtsai) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: I have the same question too. Shouldn't the ValidateSignatureLast logic applies here?

**Review by dannybtsai — APPROVED** [review]

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 8]
```diff
@@ -1221,16 +1260,17 @@ private TokenValidationResult ValidateJWE(JsonWebToken jwtToken, string decrypte
         {
             try
             {
-                var innerToken = ValidateSignature(decryptedJwt, validationParameters, configuration);
-                jwtToken.InnerToken = innerToken;
-                var innerTokenValidationResult = ValidateTokenPayload(innerToken, validationParameters, configuration);
+                TokenValidationResult tokenValidationResult = ValidateJWS(decryptedJwt, validationParameters, configuration);
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:1263]: >tokenValidationResult [](http://example.com/codeflow?start=38&length=21)

can this return null?

File: src/Microsoft.IdentityModel.Tokens/Validators.cs [hunk 9]
```diff
@@ -286,8 +282,10 @@ internal static string ValidateIssuer(string issuer, SecurityToken securityToken
                 }
             }
 
+
+
```
Comment (jennyf19) [inline src/Microsoft.IdentityModel.Tokens/Validators.cs]: nit: remove extra lines

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 10]
```diff
@@ -1004,6 +1036,68 @@ private ClaimsPrincipal ValidateToken(string token, JwtSecurityToken outerToken,
...
+            JwtSecurityToken validatedToken = ValidateSignature(token, jwtToken, validationParameters, configuration);
+
+            // use protected virtual method that does not take in configuration for back compatibility purposes
+            if (configuration == null)
+                ValidateIssuerSecurityKey(jwtToken.SigningKey, jwtToken, validationParameters);
+            else
+                Validators.ValidateIssuerSecurityKey(jwtToken.SigningKey, jwtToken, validationParameters, configuration);
+
+            return validatedToken;
+        }
```
Comment (jennyf19) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:1080]: This doesn't seem to be influenced by the above if/else?

Comment (jennyf19) [reply src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:1080]: Do the above methods throw?

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Review by RojaEnnam — APPROVED** [review]
